newsrevenuehub/rev-engine
# DEV-3152: Fix thank you share URL, migrate to TS

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

n/a

#### What's this PR do?

- Fixes a bug in PaymentSuccess where it set the incorrect URL for donationPageUrl, which caused email sharing to use the wrong URL.
- Adds better handling to GenericThankYou if you load it outside of the contribution flow; it now redirects you to the parent contribution page.
- Migrates AnalyticsContextWrapper, DonationPageFooter, and PaymentSuccess to TypeScript and adds unit tests.
    - Renames AnalyticsContextWrapper to AnalyticsContextProvider to make it match other context provider component names.
    - Refactors sharing links from GenericThankYou to a new component, PostContributionSharing.
- Replaces sharing icons with MUI icons and left-aligns all content on the thank you page.

#### Why are we doing this? How does it help us?

Fixes a bug with email sharing; continues TypeScript/Jest migration.

#### How should this be manually tested? Please include detailed step-by-step instructions.

To test the core issue:
1. Make a contribution on a page that doesn't have a custom thank you page set.
2. Confirm that the thank-you page shows the correct information (email you entered, contribution amount).
3. Click the button on the page that appears after the contribution to share via email.
4. Confirm that the URL in the email shows the contribution page URL. It should not show the thank you page URL, nor should it have any query string (e.g. `?something=something`).
5. Confirm the contribution appears in Stripe and the admin dashboard.

To test error handling of the thank you page:

1. Make a contribution on a page that doesn't have a custom thank you page set.
2. Copy and paste the URL in your browser into an incognito window.
3. This should send you to the contribution page that you started at (as opposed to showing an error).

To test custom style on thank you page for regressions:

1. Make a contribution on a page that has a custom theme, but not a custom thank you page.
2. Make a contribution via this page.
3. Confirm that the appearance of the thank you page matches the theme (fonts and colors).

To test the nonprofit message for regressions:
1. Make a contribution to a revenue program that is nonprofit, and doesn't have a custom thank you page.
2. Confirm that on the thank you page, there's a message saying that contributions are tax-deductible.
3. Repeat steps 1-2 for a revenue program that is not a nonprofit.
4. Confirm that the tax-deductible message isn't present.

To test sharing for regressions:
1. Make a contribution on a page that doesn't have a custom thank you page set.
5. Click the button on the page that appears after the contribution to share via Facebook.
6. Confirm that the URL it prompts you to share on Facebook is the the contribution page URL. It should not show the thank you page URL, nor should it have any query string (e.g. `?something=something`).
7. Click the button on the page to share via Twitter.
8. Confirm that the URL it prompts you to share on Twitter is the the contribution page URL. It should not show the thank you page URL, nor should it have any query string (e.g. `?something=something`).

To test analytics tracking for regressions:

1. Make a contribution on a page that doesn't have a custom thank you page set. The page's organization must have a Facebook pixel ID set.
5. Confirm that the confirmation created a conversion event in the page's analytics.
6. Confirm that page view events appear in analytics for the contribution page and the thank you page.
7. Make a contribution on a page that _does_ have a custom thank you page set. 
5. Confirm that page view events appear in analytics for the contribution page, but not the thank you page.

To test custom thank you redirection for regressions:

1. Make a contribution on a page that has a custom thank you page set.
3. Confirm that after the contribution, you are sent to this page. It should also have these query strings set on that page, with correct values for the contribution you just made:
   a. `frequency`
   b. `amount`
   c. `uid` - I'm not sure where this value comes from, though.
8. Confirm the contribution appears in Stripe and the admin dashboard.

To test post-contribution thank you links for regressions:

1. Make a contribution on a page that has a custom post-thank you page set.
3. Confirm that after the contribution, you see a link that reads "Return to website" and that clicking it goes to the URL set as the post-thank you page.
4. Confirm the contribution appears in Stripe and the admin dashboard.

To test the contribution page footer for regressions:

1. Go to a contribution page.
2. Confirm at the bottom of the page, it reads "(c) 2023 _revenue program name_".
4. Confirm that also at the bottom of the page, there's a link to fundjournalism.org.
6. Edit a contribution page.
7. Repeat steps 2-3 for the page preview.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

n/a

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-3152

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

No.
